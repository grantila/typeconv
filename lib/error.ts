import { codeFrameColumns } from "@babel/code-frame"
import {
	CoreTypesErrorMeta,
	locationToLineColumn,
	isCoreTypesError,
} from "core-types"


export function formatCoreTypesError( err: Error ): string
{
	if ( isCoreTypesError( err ) )
	{
		const message = `[${err.name}] ${err.message}`;
		return formatError( message, err );
	}

	return err.message;
}

export function formatError(
	message: string,
	meta?: CoreTypesErrorMeta
)
: string
{
	if ( meta?.loc?.start && meta?.source )
	{
		const loc = locationToLineColumn( meta.source, meta.loc );
		const { start, end } = loc;

		if ( start )
			return codeFrameColumns(
				meta.source,
				{ start, end },
				{ message }
			);
	}

	return message;
}
