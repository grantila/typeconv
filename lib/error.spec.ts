import { UnsupportedError } from "core-types"

import { formatError, formatCoreTypesError } from "./error.js"

describe( "error", ( ) =>
{
	it( "formatError should fallback to error message", ( ) =>
	{
		const output = formatError( "foo message" );
		expect( output ).toEqual( "foo message" );
	} );

	it( "formatError should properly display errors", ( ) =>
	{
		const output = formatError(
			"foo message",
			{
				source: "type X = bad type;",
				loc: { start: 13, end: 17 },
			}
		);
		expect( output ).toMatchSnapshot( );
	} );

	it( "formatCoreTypesError should fallback to error message", ( ) =>
	{
		const output = formatCoreTypesError( new Error( "foo error" ) );
		expect( output ).toEqual( "foo error" );
	} );

	it( "formatCoreTypesError should display nice error message", ( ) =>
	{
		const err = new UnsupportedError(
			"foo error",
			{
				source: "type X = bad type;",
				loc: { start: 13, end: 17 },
			}
		);
		const output = formatCoreTypesError( err );
		expect( output ).toMatchSnapshot( );
	} );
} );
