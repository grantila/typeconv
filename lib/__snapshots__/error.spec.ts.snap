// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error formatCoreTypesError should display nice error message 1`] = `
"> 1 | type X = bad type;
    |             ^^^^ [Error] foo error"
`;

exports[`error formatError should properly display errors 1`] = `
"> 1 | type X = bad type;
    |             ^^^^ foo message"
`;
